File: 06/06.go
package main

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
)

func findEndOfFirstMarker(input string, markerLength int) (int, error) {
	// scan through the string, character by character
	for i := range input {
		// check if all characters in our potential marker are unique
		if stringContainsUniqueChars(input[i : i+markerLength]) {
			// x unique characters!
			// return the position of the END of the marker (cursor is at the start, looking forward)
			return i + markerLength, nil
		}
	}

	return 0, errors.New("Marker never found")
}

func stringContainsUniqueChars(input string) bool {
	charSet := make(map[rune]bool)
	for _, char := range input {
		if charSet[char] {
			// we've already seen this char
			return false
		}
		charSet[char] = true
	}

	return true
}

func solve06() error {
	file, err := fs.ReadFile(os.DirFS("."), "input.txt")
	if err != nil {
		return err
	}
	input := string(file)

	endOfMarker4, err := findEndOfFirstMarker(input, 4)
	if err != nil {
		return err
	}
	fmt.Println("4", endOfMarker4)

	endOfMarker14, err := findEndOfFirstMarker(input, 14)
	if err != nil {
		return err
	}
	fmt.Println("14", endOfMarker14)

	return nil
}

func main() {
	err := solve06()
	if err != nil {
		panic(err)
	}
}

File: 06/06.rs
use std::collections::HashSet;
use std::fs;

fn find_end_of_first_marker(marker_length: usize) -> Result<usize, ()> {
    let input = fs::read_to_string("./input.txt").unwrap();

    // scan through the string, character by character
    for i in 0..input.len() {
        // look forward for a marker
        // copy the string into a hashset of it's characters to check uniqueness
        let potential_marker: HashSet<char> = input[i..i + marker_length].chars().collect();

        // check our marker is truly all unique
        if potential_marker.len() == marker_length {
            // x unique characters!
            // return the position of the END of the marker (cursor is at the start, looking forward)
            return Ok(i + marker_length);
        }
    }

    Err(())
}

fn main() {
    println!("4: {}", find_end_of_first_marker(4).unwrap());
    println!("14: {}", find_end_of_first_marker(14).unwrap());
}
